envbuilder.go

```diff
@@ -932,6 +948,41 @@ func Run(ctx context.Context, options Options) error {
 
 	logf(codersdk.LogLevelInfo, "=== Running the init command %s %+v as the %q user...", options.InitCommand, initArgs, user.Username)
 
+	execOneLifecycleScript := func(s devcontainer.LifecycleScript, name string) error {
+		if s.IsEmpty() {
+			return nil
+		}
+		logf(codersdk.LogLevelInfo, "=== Running %s as the %q user...", name, user.Username)
+		if err := s.Execute(ctx); err != nil {
+			logf(codersdk.LogLevelError, "Failed to run %s: %v", name, err)
+			return err
+		}
+		return nil
+	}
+
+	execLifecycleScripts := func() {
+		if !skippedRebuild {
+			if err := execOneLifecycleScript(scripts.OnCreateCommand, "onCreateCommand"); err != nil {
+				return
+			}
+		}
+		if err := execOneLifecycleScript(scripts.UpdateContentCommand, "updateContentCommand"); err != nil {
+			return
+		}
+		if err := execOneLifecycleScript(scripts.PostCreateCommand, "postCreateCommand"); err != nil {
+			return
+		}
+		// TODO: Technically this is supposed to run "each time the
+		// container is successfully started", so would it be more
+		// correct to pass it to InitCommand via the environment and
+		// rely on InitCommand to execute it?
```

Comment:
@kylecarbs: Curious for your thoughts on this. It's probably simplest to just execute the post start command here, but technically it's supposed to run once the container starts up successfully, which means the `InitCommand` should handle it. However, it seems burdensome to require an `InitCommand` to support this. I suppose we could make it an option?

Reply:
Interesting. I'll noodle on this for a sec.

My intuition says there maybe should be a way to manually invoke the `postStartCommand`, I'm just not sure how.

Reply:
Another alternative to passing through the `postStartCommand` through an env variable could be to create a script somewhere on the filesystem that the init command could execute once it's ready to run the `postStartCommand`.

For example, we could support a `POST_START_COMMAND_PATH` option. If you set `POST_START_COMMAND_PATH=/bin/post-start.sh`, envbuilder would create an executable script at the path `/bin/post-start.sh` that would invoke the post start commands. If `POST_START_COMMAND_PATH` is unset, envbuilder would handle the post-start commands itself prior to execing the init command.

Reply:
@kylecarbs: Does this sound like a good approach to you?

Reply:
What shell does the official devcontainer spec execute with? That's an oddity I'm not exactly sure how to fix.

I think that route is great though. I primarily want users to be able to execute the post-start command after their own start initialization, just in case they need to prepare something.

Reply:
> What shell does the official devcontainer spec execute with?

The spec says it uses `/bin/sh` to exec commands provided as strings:

> For each command property, if the value is a single string, it will be run in /bin/sh.

Reply:
> I think that route is great though. I primarily want users to be able to execute the post-start command after their own start initialization, just in case they need to prepare something.

Okay, went ahead with this. I added a `POST_START_SCRIPT_PATH` option which will create a shell script at the specified path when set, if any `postStartCommand`s were set. This allows the execution of these commands to be controlled by the init command, so it can do it after successful startup.

Reply:
Neato. It's reasonable for us to expect the same then.

It gets weird when the command is an array. What should we do in that scenario?

Reply:
> It gets weird when the command is an array. What should we do in that scenario?

I believe I implemented the proper quoting/escaping to handle this. Every arg is enclosed in single-quotes.